cassiopeia/solver/graph_utilities.py

```diff
@@ -0,0 +1,361 @@
+            ) - num / min(denom, total_weight - denom)
+
+    for u in G.nodes():
+        d = sum([G[u][v]["weight"] for v in G.neighbors(u)])
+        if d == 0:
+            return [u]
+        c = sum(
+            [
+                G[u][v]["weight"]
+                for v in G.neighbors(u)
+                if check_if_cut(u, v, new_cut)
+            ]
+        )
+        delta_n[u] = d - 2 * c
+        if u in new_cut:
+            delta_d[u] = -d
+        else:
+            delta_d[u] = d
+        set_ip(u)
+
+    all_neg = False
+    iters = 0
+
+    while (not all_neg) and (iters < len(G.nodes)):
+        best_potential = 0
+        best_index = None
+        for v in G.nodes():
+            if ip[v] < best_potential:
+                best_potential = ip[v]
+                best_index = v
+        if not best_index is None:
+            num += delta_n[best_index]
+            denom += delta_d[best_index]
+            for j in G.neighbors(best_index):
+                if check_if_cut(best_index, j, new_cut):
+                    delta_n[j] += 2 * G[best_index][j]["weight"]
+                else:
+                    delta_n[j] -= 2 * G[best_index][j]["weight"]
+                set_ip(j)
+            delta_n[best_index] = -delta_n[best_index]
+            delta_d[best_index] = -delta_d[best_index]
+            set_ip(best_index)
+            if best_index in new_cut:
+                new_cut.remove(best_index)
+            else:
+                new_cut.add(best_index)
+        else:
+            all_neg = True
+        iters += 1
+
+    return list(new_cut)
+
+
+def similarity(
+    u: int,
+    v: int,
+    cm: pd.DataFrame,
+    missing_char: str,
+    w: Optional[Dict[int, Dict[str, float]]] = None,
+) -> Union[int, float]:
+    """A function to return the number of (non-missing) character/state
+    mutations shared by two samples.
+
+    Args:
+        u: The row index of the character matrix representing the first sample
+        v: The row index of the character matrix representing the second sample
+        cm: The character matrix of observed character states for all samples
+        missing_char: The character representing missing values
+        w: A set of optional weights for edges in the connectivity graph
+    Returns:
+        The number of shared mutations between two samples, weighted or unweighted
+    """
+
+    # TODO Optimize this using masks
+    k = cm.shape[0]
+    if w is None:
+        return sum(
+            [
+                1
+                for i in range(k)
+                if cm.iloc[u, i] == cm.iloc[v, i]
+                and (cm.iloc[u, i] != "0" and cm.iloc[u, i] != missing_char)
+            ]
+        )
+    else:
+        return sum(
+            [
+                w[i][cm.iloc[u, i]]
+                for i in range(k)
+                if cm.iloc[u, i] == cm.iloc[v, i]
+                and (cm.iloc[u, i] != "0" and cm.iloc[u, i] != missing_char)
+            ]
+        )
```

Comment:
Let's take this out for now - only keep the code that's relevant for the MaxCut algorithm PR. This is hard to review without knowing how it's used in the spectral method.